ellemouton:pr21224 (#312)
bitcoin-core-review-club/website: _posts/2021-02-24-#21224.md

---
layout: pr
date: 2021-02-24
title: "Halt processing of unrequested transactions"
pr: 21224
authors: [ariard]
components: ["p2p"]
host: ellemouton
status: upcoming
commit:
---

> ## Notes

Reviewer: I think these could be a bit more detailed. For example, it wouldn't do any harm to repeat the transaction relay message flow. We've done this before, eg in the review clubs for 18238 and 15505, but it's good to repeat it for newcomers and reinforce it for others.

- Currently it is possible for peers to send each other `tx` messages even if the

>   transaction with that txid/wtxid has not been requested.

Reviewer: It's always possible for a peer to _send_ a transaction, even after this PR. The difference is that we'll then not process that transaction unless it's been requested.

- This PR aims to change that by requiring a transaction to first be requested
  by a node before it is relayed to that node from one of its peers.

- See the PR’s author's motivation behind the PR along with some of the
  concerns that this change brings about:

>   https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-February/018391.html

Reviewer: Use a markdown link for this rather than a bare url.

## Questions

> 1. Did you review the PR? Concept ACK, approach ACK, tested ACK, or NACK?

Reviewer: We usually have a link here to explain the terms. I think that's useful for first-time participants.

> 2. What is the TxRequestTracker class used for?

Reviewer: Maybe add a link to `src/txrequest.h`? Use the latest master commit for the url.

3. What are the different states that a peer/tx “announcement” can be in? Where
   in the code is an announcement shifted between the different states?

4. There are currently two ways that a node can learn about new transactions.

>    What are they?

Reviewer: I'm not sure I know what these two ways are.

5. What is the problem with the current way in which transaction messages are

>    handled?

Reviewer: Maybe "what does the author claim is the problem..." or similar. There is some disagreement that this is a problem (or that this PR solves it) https://github.com/bitcoin/bitcoin/pull/21224#issuecomment-781395303

6. What is the change that this PR is making and how does it solve the problem
   mentioned in Q5?

7. There are a lot of switches on the PR_RELAY flag in code relevant to this
   PR. What does the PF_RELAY flag mean? In what case(s) would you want to set

>    this flag for a peer?

Reviewer: Use backticks in markdown to indicate a code snippet or constant: `PR_RELAY`

8. Discussion: Could this change perhaps allow for a privacy attack, such as

>    transaction origin inference or tx-relay connection graph leaks? 

Reviewer: Do you have some answers in mind for this? If not, then such an open-ended question might not lead to very useful discussion.

Reviewer 2: See my mail reply to Jeremy, I don't think it's a real concern for this PR but I would be glad having reviewers thinking twice on it. The TxProbe paper is a good one for anyone willingly to dig further.

> 9. Discussion: Since the INV -> GETDATA -> TX sequence has not been necessary

Reviewer: I suggest lowercase for message types (e.g. `inv` -> `getdata` -> `tx`), since that's what they're serialized as in the p2p messages.

   for communicating and receiving transaction data, some other clients don't
   bother with the sequence at all. This means that  if this change was deployed

>    to Core nodes then other clients would not be able to relay transaction

Reviewer: s/Core/Bitcoin Core/

>    messages and so they will have to adapt and update. Do the pros out weigh these

Reviewer: maybe s/relay transaction messages/relay transactions/

     request the full transaction.
  3. the relaying node delivers a `tx` message to the receiving node. If the
     relaying node is no longer able to deliver the transaction, it responds
     with `notfound` instead of the `tx`.

  You can learn more about these messages [here](https://en.bitcoin.it/wiki/Protocol_documentation#inv).

- Currently a node will process a transaction that it receives even if it has not 
  made a request for that transaction (i.e: If it recieves a `tx` message without
  having first sent a `getdata` message asking for the `tx` message then it will
  still process the transaction).

- This PR aims to change that by requiring a transaction to first be requested
  by a node before it is relayed to that node from one of its peers.

- See the PR’s author's motivation behind the PR along with some of the
  concerns that this change brings about:
  [see mailing list thread here](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-February/018391.html)

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
   NACK?](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)

> 2. What is the `TxRequestTracker` class used for? See [`src/txrequest.h`](https://github.com/bitcoin/bitcoin/blob/b805dbb0b9c90dadef0424e5b3bf86ac308e103e/src/txrequest.h)

Reviewer 2: IIRC the invblock paper is referred in `TxRequestTracker` doc, not the TxProbe one which is also super relevant for tx-relay : https://arxiv.org/pdf/1812.00942.pdf

3. What are the different states that a peer/tx “announcement” can be in? Where
   in the code is an announcement shifted between the different states?

4. What does the author claim that the problem is with the current way in which 
   transaction messages are handled?

5. What is the change that this PR is making and how does it solve the problem
   mentioned in Q5?

> 6. There are a lot of switches on the `PR_RELAY` flag in code relevant to this

Reviewer 2: I think you should raise up the question if this PR should introduce a new permission to unify tx-announcement related behaviors which could be  different from link-type authorizations. At least I'm waiting feedback on this.

---
layout: pr
date: 2021-02-24
title: "Halt processing of unrequested transactions"
pr: 21224
authors: [ariard]
components: ["p2p"]
host: ellemouton
status: upcoming
commit:
---

## Notes

- Transaction relay is a three step process: `inv` -> `getdata` -> `tx`:

  1. the relaying node sends an `inv` message to the receiving node to announce
     a new transaction.

>   2. the receiving node sends a `getdata` message to the relaying node to

Reviewer: Perhaps add something like "if the receiving node wants to retrieve the transaction..." or similar. The receiving node won't request transactions that it's seen before, and will only request from one peer if it receives invs for the same transaction from multiple peers.

     request the full transaction.
  3. the relaying node delivers a `tx` message to the receiving node. If the
     relaying node is no longer able to deliver the transaction, it responds
     with `notfound` instead of the `tx`.

  You can learn more about these messages [here](https://en.bitcoin.it/wiki/Protocol_documentation#inv).

- Currently if a `tx` message is received it will not be processed if:

> 	1. The node is in blocks-only mode and the peer sending the message has 

Reviewer: can you replace tabs with spaces please :pray:

	no relay permissions
	2. The peer is a blocks-relay only peer

>   In all other cases, the node will process a transaction that it receives even 

Reviewer: It depends what you mean by 'process' here. We'll do the `AlreadyHaveTx()` checks first, to make sure that we don't process the same transaction more than once: https://github.com/bitcoin/bitcoin/blob/9f3ffa29389aa59ea87b5aa974a2da17ccfd31a4/src/net_processing.cpp#L1619-L1648.

  if it has not made a request for that transaction (i.e: If it recieves a `tx` 
  message without having first sent a `getdata` message asking for the `tx` message 
  then it will still process the transaction). In other words the 3 step process
  described above is not enforced.

- This PR aims to change that by requiring a transaction to first be requested
  by a node before it processes an accociated `tx` message.

> - See the PR’s author's motivation behind the PR along with some of the
>   concerns that this change brings about:
>   [see mailing list thread here](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-February/018391.html)

Reviewer: ```suggestion
- The PR author provides additional motivation for the PR and raises some of the
  concerns that this change brings about in a [mailing list thread](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-February/018391.html).
```

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
   NACK?](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)

> 2. What is the `TxRequestTracker` class used for? See [`src/txrequest.h`](https://github.com/bitcoin/bitcoin/blob/b805dbb0b9c90dadef0424e5b3bf86ac308e103e/src/txrequest.h)

Reviewer: ```suggestion
2. What is the `TxRequestTracker` class in [src/txrequest.h](https://github.com/bitcoin/bitcoin/blob/b805dbb0b9c90dadef0424e5b3bf86ac308e103e/src/txrequest.h) used for?
```

3. What are the different states that a peer/tx “announcement” can be in? Where
   in the code is an announcement shifted between the different states?

4. What does the author claim that the problem is with the current way in which 
   transaction messages are handled?

5. What is the change that this PR is making and how does it solve the problem
   mentioned in Q4?

> 6. There are a lot of switches on the `PR_RELAY` flag in code relevant to this

Reviewer: typo: PR_RELAY -> PF_RELAY

   PR. What does the `PF_RELAY` flag mean? In what case(s) would you want to set
   this flag for a peer?

7. Discussion: Since the `inv` -> `getdata` -> `tx` sequence has not been necessary
   for communicating and receiving transaction data, some other clients don't
   bother with the sequence at all. This means that  if this change was deployed
   to Bitcoin Core nodes then other clients would not be able to relay transactions 

>    and so they will have to adapt and update. Do the pros out weigh these

Reviewer: "not be able to relay transactions to upgraded Bitcoin Core nodes. Eventually upgraded nodes would make up the majority of the network, and so those clients would have to ..."
